src/util/event_emitter.js

```diff
@@ -0,0 +1,140 @@
+export class EventEmitter {
+  constructor() {
+    this._handlers = [];
+  }
+
+  /**
+   * Adds the event name and handler function to the end of the handlers array.
+   * No checks are made to see if the handler has already been added.
+   * Multiple calls passing the same combination of event name and handler will result in the handler being added,
+   * and called, multiple times.
+   * @param {String} event
+   * @param {Function} handler
+   * @returns {EventEmitter}
+   */
+  on(event, handler) {
+    this._handlers.push({
+      event,
+      handler
+    });
+
+    return this;
+  }
+
+  _onceWrapper(...args) {
+    if (!this.fired) {
+      this.target.off(this.event, this.wrapFn);
+      this.fired = true;
+      Function.prototype.apply.call(this.handler, this.target, [...args]);
+    }
+  }
+
+  _onceWrap(event, handler) {
+    const state = {
+      fired: false,
+      wrapFn: undefined,
+      target: this,
+      event,
+      handler
+    };
+    const wrapped = this._onceWrapper.bind(state);
+    wrapped.handler = handler;
+    state.wrapFn = wrapped;
+    return wrapped;
+  }
+
+  /**
+   * Adds a one-time handler function for the named event.
+   * The next time event is triggered, this handler is removed and then invoked.
+   * @param {String} event
+   * @param {Function} handler
+   * @returns {EventEmitter}
+   */
+  once(event, handler) {
+    return this.on(event, this._onceWrap(event, handler));
+  }
+
+  /**
+   * Removes all instances for the specified handler from the handler array for the named event.
+   * @param {String} event
+   * @param {Function} handler
+   * @returns {EventEmitter}
+   */
+  off(event, handler) {
+    this._handlers = this._handlers.filter(item => {
+      return item.event !== event || item.handler !== handler;
+    });
+
+    return this;
+  }
+
+  /**
+   * Synchronously calls each of the handlers registered for the named event,
+   * in the order they were registered, passing the supplied arguments to each.
+   * @param {String} event
+   * @param  {any[]} args
+   * @returns {Boolean} true if the event had handlers, false otherwise.
+   */
+  emit(event, ...args) {
+    let called = false;
+    this._handlers.forEach(item => {
+      if (typeof item.handler === 'function') {
```

Comment:
What about do this when we add the handler ? it's probably better no ?

Reply:
I added checks in `on` function and removed the check here